Refactor `Binding` data structures in preparation for partition

Keno opened the pull request:
This is a re-worked extraction of #54654, adjusted to support the new semantics arrived at over the course of that thread. Note that this is the data-structure change only. The semantics in this PR still match current master to the greatest extent possible.

The core idea here is to split `Binding` in two: A new `Binding` with minimal data and a `BindingPartition` that holds all data that is world-age partitioned. In the present PR, these are always in 1:1 correspondednce, but after #54654, there will be multiple `BindingPartition`s for every `Binding`.

Essentially the `owner` and `ty` fields have been merged into a new `restriction` field of `BindingPartition`, which may also hold the value of a constant (consistent with the final semantics reached in #54654). The non-partitioned binding->value field is now used exclusively for non-constant globals. The disambiguation for how to interpret the `restriction` field happens via flags. `->imported` grew to 2 bits and can now be one of `NONE`/`IMPLICIT`/ `EXPLICIT`/`GUARD`. `GUARD` corresponds to the old `b->owner == NULL` case. `NONE` corresponds to the old `b->owner == b` case, while IMPLICIT/EXPLICIT correspond to `b->owner != b` and the old `imported` flag. Other than that, the behavior of the flags is unchanged.

Additionally, fields are provided for `min_world`/`max_world`/`next`, but these are unused in this PR and prepratory only.

topolarity commented on src/codegen.cpp:
```diff
@@ -3234,8 +3252,9 @@ static jl_cgval_t emit_globalop(jl_codectx_t &ctx, jl_module_t *mod, jl_sym_t *s
+    jl_binding_partition_t *bpart = jl_get_binding_partition(bnd, ctx.max_world);
+    if (bpart && !bpart->constp) {
+        jl_value_t *ty = bpart->restriction;
         if (ty != nullptr) {
```
I believe this needs additional guards now, for this case:
```julia
julia> global x::Int64
julia> foo() = (global x; x)
julia> foo() # emits code w/ stores/loads directly to binding->value
julia> const x = 15 # does not update binding->value any more
julia> foo() # should be 15
```

Keno replied:
Invalidation will take care of this.

Keno commented on src/julia.h:
```diff
@@ -627,19 +627,47 @@ typedef struct _jl_weakref_t {
+     * First look at ->imported (see above). If NONE, but `->constp`, this holds the
+     * value of the constant. Otherwise this is a global and holds the type restriction.
+     */
+    jl_value_t *restriction;
```
I think that would be semantically consistent and reasonable, but there isn't really any benefit to doing so.

Keno commented on src/julia.h:
```diff
@@ -627,19 +627,47 @@ typedef struct _jl_weakref_t {
+    //  ->restriction does not hold any import-related data.
+    BINDING_IMPORT_NONE     = 0x0,
+    // Implicit: The binding was implicitly import from a `using`'d module.
+    //  ->restriction holds the imported binding
```
Yes, it short circuits. I haven't worked out the backedges yet.

Keno commented on src/julia.h:
```diff
@@ -627,19 +627,47 @@ typedef struct _jl_weakref_t {
     uint8_t publicp:1; // exportp without publicp is not allowed.
-    uint8_t imported:1;
+    uint8_t imported:2; // enum jl_binding_import
     uint8_t usingfailed:1;
```
Yes

Keno committed:
Refactor binding flags

Keno committed:
Join global ty and constant value into binding->restriction

Keno committed:
Merge owner field into restriction also

Keno committed:
Add another kind of constant binding

`import A` is a bit of a weird beast. We want to treat it as an import
for reflection purposes, but internally, it looks like a constant.
After some discussion, it really is just its own kind of thing, so
give it a special binding kind.

Keno committed:
Renamed ->imported ->kind

Keno committed:
Split Binding

Keno commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54788#issuecomment-2192662294) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c2cc636_vs_14956a1/report.html) is available.

Keno committed:
Remove dual import assertion

Keno commented:
@nanosoldier `runtests(["Tricks", "ComputationalResources", "ModuleInterfaceTools", "MultilineStrings", "ImportAll", "ForceImport", "InvariantsCore", "Syslogs", "TranscodingStreams", "EulerAngles", "MarkdownAST", "MD5", "StrFormat", "MiniLoggers", "JACC", "Invariants", "Automa", "BufferedStreams", "Setfield", "ClimaUtilities", "OctreeBH", "NonNegLeastSquaresMLJInterface", "HolidayCalendars", "KiwiConstraintSolver", "MixedStructTypes", "DistributedArrays", "MLJTSVDInterface", "HMMER", "CryptographicHashFunctions", "FASTX", "Neighborhood", "ShuffleProofs", "HOHQMesh", "Rocket", "JSON3", "TypedFASTX", "StrEntities", "OutlierDetectionTrees", "QuantumAlgebra", "SelfOrganizingMaps", "OneRule", "MLJDecisionTreeInterface", "LinearFold", "PAndQ", "ArtGallery", "OutlierDetectionNeighbors", "PkgJogger", "MultivariateMoments", "OnlineTechnicalIndicators", "ParallelKMeans", "ModuleInfo", "CharacteristicInvFourier", "MetaGraphsNext", "Strs", "BioRecordsProcessing", "GFF3", "ToolipsSession", "DataGraphs", "GAP", "BurrowsWheelerAligner", "TerminalUserInterfaces", "MLJClusteringInterface", "PreprocessMD", "InferenceObjects", "BibParser", "LazyAlgebra", "Bibliography", "CitationRecipes", "ChunkedCSV", "Infernal", "Logomaker", "NLopt", "CUBScout", "BlackBoxOptim", "MLJGLMInterface", "ExtendableGrids", "AbstractPPL", "StateSpaceEcon", "StaticWebPages", "DocumenterCitations", "KroneckerProductKernels", "Microbiome", "AbstractLogic", "Gaugefields", "COBREXA", "ScikitLearn", "BaytesOptim", "Bijectors", "Gtk", "DoseCalculators", "MLJLinearModels", "TwoDots", "MEstimation", "TuringCallbacks", "GtkSourceWidget", "MLJNaiveBayesInterface", "StatsLearnModels", "SliceSampling", "Ripserer", "MLJXGBoostInterface", "Baytes", "PhyloCoalSimulations", "DrugInteractions", "NonconvexMMA", "ClimaCoreSpectra", "MetidaNLopt", "QuartetNetworkGoodnessFit", "ProfileView", "ClimaCoreVTK", "OutlierDetectionNetworks", "TwoDotsModels", "PsychomotorVigilanceTask", "MLJEnsembles", "PhyloPlots", "MLJText", "MGVI", "MLJTestInterface", "OutlierDetectionTest", "MLJIteration", "ClimaCoreTempestRemap", "OutlierDetection", "PhyloGaussianBeliefProp", "RandomFeatures", "MLJParticleSwarmOptimization", "MCMCTempering", "YAXArrays", "MLJSerialization", "SpikeSorting", "MLJTuning", "ITensorGaussianMPS", "GenerativeTopographicMapping", "TenNetLib", "MicroCanonicalHMC", "EvoLinear", "PhyloDiamond", "GMMParameterEstimation", "PowerPlots", "ITensorVisualizationBase", "MAGEMinApp", "DecomposingPolynomialSystems", "MLJTestIntegration", "JointEnergyModels", "TrajGWAS", "MLJ", "BellDiagonalQudits", "GameTheory", "NonconvexJuniper", "NonconvexIpopt", "NonconvexPavito", "EqualitySampler", "RFFMakie", "PolynomialGTM", "SeisMakie", "GtkUtilities", "SimpleCrop", "SpectralStatistics", "PhyNEST", "ProbabilisticEchoInversion", "LeafGasExchange", "Garlic", "TaijaData", "ObservableCortex", "DiskArrayEngine", "MendelImpute", "SharedArrays", "QuadraticProgramNetworks"])`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54788#issuecomment-2194167482) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/6282a92_vs_14956a1/report.html) is available.

Keno committed:
Fix importing over a previously defined constant

Keno commented:
@nanosoldier `runtests(["ModuleInterfaceTools", "Tricks", "Syslogs", "ComputationalResources", "JACC", "OctreeBH", "MarkdownAST", "Invariants", "StrFormat", "InvariantsCore", "HolidayCalendars", "KiwiConstraintSolver", "Neighborhood", "MiniLoggers", "FASTX", "ArtGallery", "CryptographicHashFunctions", "Automa", "TypedFASTX", "MultilineStrings", "ShuffleProofs", "ToolipsSession", "MD5", "TranscodingStreams", "TerminalUserInterfaces", "MixedStructTypes", "BurrowsWheelerAligner", "BioRecordsProcessing", "CitationRecipes", "ModuleInfo", "BibParser", "CUBScout", "AbstractPPL", "OutlierDetectionTrees", "LinearFold", "GFF3", "AbstractLogic", "Bibliography", "StaticWebPages", "PAndQ", "Strs", "Gtk", "DoseCalculators", "Setfield", "QuantumAlgebra", "TwoDots", "PkgJogger", "DataGraphs", "GtkSourceWidget", "OneRule", "NonNegLeastSquaresMLJInterface", "BlackBoxOptim", "MLJClusteringInterface", "MEstimation", "MLJTSVDInterface", "DrugInteractions", "Microbiome", "KroneckerProductKernels", "PreprocessMD", "PsychomotorVigilanceTask", "MetaGraphsNext", "TwoDotsModels", "ParallelKMeans", "Rocket", "Logomaker", "OutlierDetectionNeighbors", "ProfileView", "MLJTestInterface", "BaytesOptim", "MLJDecisionTreeInterface", "SelfOrganizingMaps", "HMMER", "MLJText", "DocumenterCitations", "MLJNaiveBayesInterface", "SpikeSorting", "MultivariateMoments", "OutlierDetectionTest", "MLJSerialization", "DistributedArrays", "Infernal", "PhyloCoalSimulations", "PhyloPlots", "ChunkedCSV", "MLJXGBoostInterface", "MLJGLMInterface", "QuartetNetworkGoodnessFit", "ExtendableGrids", "HOHQMesh", "StateSpaceEcon", "MLJIteration", "MLJParticleSwarmOptimization", "MLJTestIntegration", "Gaugefields", "SliceSampling", "MGVI", "MetidaNLopt", "OutlierDetection", "NLopt", "GMMParameterEstimation", "InferenceObjects", "MLJEnsembles", "SharedArrays", "DecomposingPolynomialSystems", "GtkUtilities", "COBREXA", "PhyloDiamond", "ScikitLearn", "ITensorGaussianMPS", "Bijectors", "MLJLinearModels", "EvoLinear", "PhyloGaussianBeliefProp", "JSON3", "NonconvexMMA", "YAXArrays", "BellDiagonalQudits", "MAGEMinApp", "MLJTuning", "LazyAlgebra", "StatsLearnModels", "TuringCallbacks", "OutlierDetectionNetworks", "GenerativeTopographicMapping", "TenNetLib", "ITensorVisualizationBase", "Ripserer", "MCMCTempering", "PhyNEST", "SeisMakie", "RandomFeatures", "JointEnergyModels", "EqualitySampler", "TaijaData", "ObservableCortex", "DiskArrayEngine", "SpectralStatistics", "NonconvexPavito", "MendelImpute", "LeafGasExchange", "RFFMakie", "PowerPlots", "SimpleCrop", "MLJ", "TrajGWAS", "Garlic", "GameTheory", "NonconvexJuniper", "NonconvexIpopt"])`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54788#issuecomment-2194695344) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/aa4b302_vs_14956a1/report.html) is available.

Keno committed:
Forward port rebase

Keno committed:
Fix tests

Keno committed:
Fix and add test for the Gtk pattern

Keno commented:
@nanosoldier `runtests(["Tricks", "ComputationalResources", "Syslogs", "JACC", "OctreeBH", "Neighborhood", "ToolipsSession", "Automa", "MixedStructTypes", "AbstractLogic", "TwoDots", "DoseCalculators", "GtkSourceWidget", "QuantumAlgebra", "Gtk", "PkgJogger", "TwoDotsModels", "MEstimation", "PsychomotorVigilanceTask", "DrugInteractions", "BlackBoxOptim", "Rocket", "ProfileView", "SpikeSorting", "ExtendableGrids", "MultivariateMoments", "DistributedArrays", "StateSpaceEcon", "MGVI", "SharedArrays", "GtkUtilities", "NonconvexMMA", "NLopt", "MetidaNLopt", "Gaugefields", "ScikitLearn", "LazyAlgebra", "JSON3", "RandomFeatures", "MLJTuning", "COBREXA", "BellDiagonalQudits", "TrajGWAS", "EqualitySampler", "PowerPlots", "MendelImpute", "GameTheory"])`

Keno committed:
Restore old module replacement logic

Keno committed:
Restore old behavior for const reassign

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54788#issuecomment-2267275041) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/5443fcd_vs_f2f188d/report.html) is available.

Keno committed:
Add atomic annotations

This won't be required for the final version, but the semantics-preserving
version updates these in places, so this needs to be atomic.

Keno committed:
Try to find the correct extern inline definition for windows

Keno commented:
@nanosoldier `runtests(["Tricks", "Syslogs", "ComputationalResources", "JACC", "OctreeBH", "Neighborhood", "ToolipsSession", "Gtk", "DoseCalculators", "AbstractLogic", "Automa", "SharedArrays", "MEstimation", "MixedStructTypes", "DrugInteractions", "TwoDotsModels", "PsychomotorVigilanceTask", "SpikeSorting", "ProfileView", "BlackBoxOptim", "GtkUtilities", "PkgJogger", "MetidaNLopt", "NonconvexMMA", "MGVI", "Rocket", "StateSpaceEcon", "ScikitLearn", "BellDiagonalQudits", "Gaugefields", "NLopt", "COBREXA", "DistributedArrays", "PowerPlots", "TrajGWAS", "MultivariateMoments", "GameTheory", "ExtendableGrids", "MLJTuning", "EqualitySampler", "RandomFeatures", "LazyAlgebra", "MendelImpute", "JSON3", "QuantumAlgebra"])`

Keno committed:
typo fixes

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54788#issuecomment-2267349557) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/27286c2_vs_f2f188d/report.html) is available.

Keno committed:
Fix module replacement

Keno committed:
Restore ability to setglobal! declared binding

Keno commented:
@nanosoldier `runtests(["Tricks", "Syslogs", "ComputationalResources", "JACC", "OctreeBH", "ToolipsSession", "SharedArrays", "MixedStructTypes", "PkgJogger", "MetidaNLopt", "NonconvexMMA", "BellDiagonalQudits", "ScikitLearn", "Gaugefields", "NLopt", "COBREXA", "DistributedArrays", "RandomFeatures", "GameTheory", "ExtendableGrids", "TrajGWAS", "PowerPlots", "MendelImpute", "JSON3", "QuantumAlgebra"])`

Keno committed:
Tweak reassignment corner cases

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54788#issuecomment-2278940608) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/14bd971_vs_f2f188d/report.html) is available.

Keno commented:
Alright, I've restored all the old semantics, with the exception of
```
const x = 2
x = 3
```
and other attempts to redefine constants using setglobal!, which is now an error (and will be after full partition). Constant redefinition using the keyword, i.e.
```
const x = 2
const x = 3
```
and remains undefined behavior (for now).

I'm planning to address the atomics concerns and then merge this in the near future.

Keno committed:
Fully initialize Binding object before allocating partition

Keno committed:
Address atomicness concerns

Current julia master changes binding kinds implicitly based
on the status of the owner field and does so atomically.
The fully partitioned version of this will do so using atomic
operations on the partition list.

This current state is a bit of a weird hybrid where we want
to change both the ->kind and the ->restriction field. To allow
this to happen, merge them into one field that has custom GC
support to ignore the low bits.

Keno commented:
If this passes CI, it should be good to go.

Keno commented:
Welp, guess I was wrong about how many bits we have. That said, I think it's fine to use a doubleword atomic for 32bit. We enforce those for all architectures while 128 bit atomics are technically still optional.

Keno committed:
Use doublword atomics on 32bit

Keno committed:
Silence analyzegc

Keno committed:
small fixes

Keno committed:
Adjust binding_kind

Keno committed:
Explain new BindingPartition to heap snapshot

Keno committed:
Force 8-byte alignment for binding partition

Keno committed:
Fix syntax test

Keno committed:
Fix Revise

Keno commented:
Good to go I think. The code to make it behave like the old thing is not particularly pretty, but should be as correct as the old code and be ok for a few weeks. I'll plan to merge this shortly.

vtjnash commented on base/client.jl:
```diff
@@ -532,6 +532,10 @@ The thrown errors are collected in a stack of exceptions.
 """
 global err = nothing
 
+# For numbered REPL prompt
+global Out::Any
+export Out
+
```
This doesn't seem like it should be here

Keno replied:
It reserves the binding. With me putting the old semantics back, it's probably no longer required, but may as well do it now.

Keno replied:
(Note that this is inside `MainInclude` like `ans` and `err`.

vtjnash replied:
IIUC, this is not supposed to have reserved the binding, as this binding actually is not supposed to be part of MainInclude originally, but rather is part of a plugin that sometimes adds this with this name for some alternative REPLs (but isn't necessarily the only name for it)

Keno replied:
Adding the binding at some random unspecified later time is just annoying though.

vtjnash replied:
That is sort of the point though, that although it is annoying to implement, but it is required to work correctly

vtjnash commented on base/reflection.jl:
```diff
@@ -207,6 +207,23 @@ function _fieldnames(@nospecialize t)
+end
+
+binding_kind(bpart::Core.BindingPartition) = ccall(:jl_bpart_get_kind, UInt8, (Any,), bpart)
+binding_kind(m::Module, s::Symbol) = binding_kind(lookup_binding_partition(get_world_counter(), GlobalRef(m, s)))
```
`get_world_counter` is usually not the correct world for queries, since it is not inferable or relevant info for the caller

Keno replied:
World argument is unused in this branch, but I can switch to `tls_world_age` if you think that makes more sense.  The only user of this on this branch is the REPL reflection code. In fully partitioned version of this, all of that code will need to become world age aware.

vtjnash commented on base/show.jl:
```diff
@@ -1021,10 +1021,14 @@ end
-    from_owner = ccall(:jl_binding_owner, Ptr{Cvoid}, (Any, Any), from, sym)
+    isdeprecated(parent, sym) && return false
+    if isconst(from, sym) && isconst(parent, sym)
+        isdefined(from, sym) || return false
```
Also need to check that `isdefined(parent, sym)` here. But is it really necessary now to do all of these lookups separately, instead of before where it could optimize doing the lookup of the Binding once and then doing all of the queries against that?

Keno replied:
I don't believe this change is necessary without the full partition, so I can revert it for now.

vtjnash commented on doc/src/manual/variables-and-scoping.md:
```diff
@@ -743,7 +743,7 @@ ERROR: invalid redefinition of constant x
 
-julia> y = 2.0
+julia> const y = 2.0
 WARNING: redefinition of constant y. This may fail, cause incorrect answers, or produce other errors.
```
Shouldn't we still expect this warning to be emitted when redefining constants, so this test change is not really necessary?

Keno replied:
See above, this is the one semantic change I kept, I made
```
const x = 2
x = 2
```
and error without `const` keyword. Same applies to `setglobal!` in general. It's no longer trivial to support in codegen because these are stored in different places now and it just didn't seem worth the extra effort only to rip it out again.

vtjnash replied:
okay, yes, upgrading a warning -> error here sounds fine to me, even though it necessitated changing the tests

vtjnash commented on src/ast.c:
```diff
@@ -204,8 +205,14 @@ static value_t fl_nothrow_julia_global(fl_context_t *fl_ctx, value_t *args, uint
     jl_binding_t *b = jl_get_module_binding(mod, var, 0);
-    b = b ? jl_atomic_load_relaxed(&b->owner) : NULL;
-    return b != NULL && jl_atomic_load_relaxed(&b->value) != NULL ? fl_ctx->T : fl_ctx->F;
+    jl_binding_partition_t *bpart = jl_get_binding_partition(b, jl_current_task->world_age);
+    ptr_kind_union_t pku = jl_walk_binding_inplace(&b, &bpart, jl_current_task->world_age);
```
This query question is occurring in `jl_get_world_counter()` (or a future value), not in `world_age`

Keno replied:
It's irrelevant in this PR, but in the full partition, this query will be for all world ages (which is not a function we have yet) and invalidation will invalidate the assumption if it changes.

vtjnash commented on src/builtins.c:
```diff
@@ -1369,19 +1369,10 @@ JL_CALLABLE(jl_f_get_binding_type)
-    return ty;
+    jl_value_t *ret = jl_get_binding_type(mod, var);
+    if (ret == jl_nothing)
+        return (jl_value_t*)jl_any_type;
```
This is may be introducing unsoundness if `jl_get_binding_type` can later change return value, since it is deleting the code that protected against that unsoundness from occuring

Keno replied:
The `old_ty == NULL` state doesn't exist anymore and without partition, the binding type cannot change.

vtjnash commented on src/interpreter.c:
```diff
@@ -94,9 +94,7 @@ static jl_value_t *eval_methoddef(jl_expr_t *ex, interpreter_state *s)
-        _Atomic(jl_value_t*) *bp = &b->value;
-        jl_value_t *gf = jl_generic_function_def(fname, modu, bp, b);
-        return gf;
+        return jl_declare_const_gf(b);
```
It is typically better to pass `modu` and `fname` here explicitly, rather than relying on extracting something from `b` and hoping it is relevant, as they aren't necessarily identical to the value in `b` if it was imported with a different name (or in the future, some Binding might not have a name or module associated at all)

Keno replied:
I'll change that in the future. It's a little awkward right now with the way that binding lookup is done in codegen.

vtjnash replied:
What was wrong with keeping the existing arguments in place here?

vtjnash commented on src/jltypes.c:
```diff
@@ -3147,12 +3147,23 @@ void jl_init_types(void) JL_GC_DISABLED
+                        jl_svec(5, jl_uint64_type /* Special GC-supported union of Any and flags*/,
+                        jl_ulong_type, jl_ulong_type, jl_any_type/*jl_binding_partition_type*/, jl_ulong_type),
+                        jl_emptysvec, 0, 1, 0);
+    const static uint32_t binding_partition_atomicfields[] = { 0x000d }; // Set fields 1, 3, 4 as atomic
```
We've been commonly changing these to bit values, since C23 now requires compilers to support those:
```suggestion
    const static uint32_t binding_partition_atomicfields[] = { 0b01101 }; // Set fields 1, 3, 4 as atomic
```
https://en.cppreference.com/w/c/language/integer_constant

vtjnash commented on src/jltypes.c:
```diff
@@ -3147,12 +3147,23 @@ void jl_init_types(void) JL_GC_DISABLED
+                        jl_emptysvec, 0, 1, 0);
+    const static uint32_t binding_partition_atomicfields[] = { 0x000d }; // Set fields 1, 3, 4 as atomic
+    jl_binding_partition_type->name->atomicfields = binding_partition_atomicfields;
+    const static uint32_t binding_partition_constfields[] = { 0x0002 }; // Set field 2 as constant
```
min_world should not be declared constant due to the serialization semantics, and the eagerness of our compiler to hoist anything it sees declared as constant

Keno replied:
Fair point

vtjnash commented on src/julia.h:
```diff
@@ -611,19 +611,73 @@ typedef struct _jl_weakref_t {
+    // Explicit: The binding was explicitly `using`'d by name
+    //  ->restriction holds the imported binding
+    BINDING_KIND_EXPLICIT     = 0x4,
+    // Imported: The binding was explicitly `using`'d by name
```
```suggestion
    // Imported: The binding was explicitly `import`'d by name
```

vtjnash commented on src/julia.h:
```diff
@@ -611,19 +611,73 @@ typedef struct _jl_weakref_t {
+    //  ->restriction is NULL.
+    BINDING_KIND_FAILED       = 0x6,
+    // Declared: The binding was declared using `global` or similar
+    // Failed: We attempted to import the binding, but the import was ambiguous
+    //  ->restriction is NULL.
```
```suggestion
```

vtjnash commented on src/julia_internal.h:
```diff
@@ -859,6 +859,92 @@ jl_method_t *jl_make_opaque_closure_method(jl_module_t *module, jl_value_t *name
+}
+
+EXTERN_INLINE_DECLARE jl_binding_partition_t *jl_get_binding_partition(jl_binding_t *b, size_t world) JL_NOTSAFEPOINT {
+    if (b && jl_is_globalref(b))
+        b = ((jl_globalref_t*)b)->binding;
```
Why would a binding be a globalref-type object?

Keno replied:
For ease of reflection, this allows both. I can update the signature.

vtjnash replied:
Seems like the caller should be aware of which?

Keno replied:
I can clean it up

vtjnash commented on src/julia_internal.h:
```diff
@@ -859,6 +859,92 @@ jl_method_t *jl_make_opaque_closure_method(jl_module_t *module, jl_value_t *name
+    while (1) {
+        if (!*bpart)
+            return encode_restriction(NULL, BINDING_KIND_GUARD);
+        ptr_kind_union_t pku = jl_atomic_load_acquire(&(*bpart)->restriction);
```
What is this load acquiring?

Keno replied:
Needs to be sure the binding it's dereferencing through has been fully initialized.

vtjnash replied:
We typically assume that relaxed loads actually give consume ordering, as long as you aren't running julia on alpha

Keno replied:
Ok, I can change it to relaxed.

vtjnash commented on src/method.c:
```diff
@@ -215,6 +215,28 @@ static jl_value_t *resolve_globals(jl_value_t *expr, jl_module_t *module, jl_sve
             if (e->head == jl_method_sym || e->head == jl_module_sym || e->head == jl_throw_undef_if_not_sym) {
                 i++;
             }
+            if (e->head == jl_assign_sym && binding_effects) {
```
Why is this allocating bindings if lowering told it not to (by omitting the `'(global ,lhs)` expr)?

Keno replied:
The intended design was that `'(global ,lhs)` affects scoping, but whether or not the presence of a GlobalRef allocates a mutable, global variable (as opposed to referencing an existing constant) depends on whether there's an assignment.

vtjnash replied:
That sounds somewhat breaking though, and a separate question to resolve later? As existing, we have it intended that `global` makes it a module global, regardless of what kind of assignment might occur later to it (so that you can potentially pre-reserve some names to prevent them from getting imported via `using` before you have a chance to assign them a value)
```
julia> f() = (global sin; sin)
f (generic function with 1 method)

julia> sin
ERROR: UndefVarError: `sin` not defined in `Main`
Suggestion: add an appropriate import or assignment. This global was declared but not assigned.
Hint: a global variable of this name also exists in Base.

julia> f() = cos
f (generic function with 2 methods)

julia> cos
cos (generic function with 14 methods)
```

vtjnash replied:
But also note that this IR construct `global` is separate and distinct from the AST construct `global`, and there is a pretty significant argument that we should give them different names. But whereas the later is a scope construct, the latter is toplevel instruction from lowering to allocate a global

Keno replied:
Yes, I want to change that, but that change is not in this PR. It's quite possible that this change is not required in this PR - I can try dropping it.

vtjnash commented on src/method.c:
```diff
@@ -215,6 +215,28 @@ static jl_value_t *resolve_globals(jl_value_t *expr, jl_module_t *module, jl_sve
+                    if (mod == module) {
+                        // Assignment does not create bindings in foreign modules (#54678)
+                        jl_binding_t *b = jl_get_module_binding(mod, name, 1);
+                        jl_binding_partition_t *bpart = jl_get_binding_partition(b, jl_current_task->world_age);
```
It may be unlikely that `jl_current_task->world_age` is correct here, since this runs in the toplevel context (constantly changing worlds) not in a fixed world

Keno replied:
This entire code section needs to be different for actual partition - in this PR, the value is unused.

vtjnash commented on src/staticdata.c:
```diff
@@ -922,14 +937,17 @@ static void jl_insert_into_serialization_queue(jl_serializer_state *s, jl_value_
     else if (jl_typetagis(v, jl_module_tag << 4)) {
         jl_queue_module_for_serialization(s, (jl_module_t*)v);
     }
+    else if (jl_is_binding_partition(v)) {
+        jl_binding_partition_t *bpart = (jl_binding_partition_t*)v;
+        jl_queue_for_serialization_(s, decode_restriction_value(jl_atomic_load_relaxed(&bpart->restriction)), 1, immediate);
+        jl_queue_for_serialization_(s, get_replaceable_field((jl_value_t**)&bpart->next, 0), 1, immediate);
```
IIUC, this will not serialize/deserialize correctly if the user (or inference) creates a reference to it. We need to determine how to correctly reattach them to the Binding / module that owns them (maybe by recording the location of all BindingPartitions in the system, as a pre-processing pass, so we can query here the Binding that owns it?)

Keno replied:
What's the particular issue you're imagining? Currently they're attached 1:1. In the full thing, you're not allowed to create new partitions for a binding already in the image, so they don't need to be ported.

vtjnash replied:
Ah right, the code talks about "walking" the list of them. But it is currently always guaranteed to actually just be a single value that gets mutated in place (e.g. when a `using` finally gets resolved)?

Keno replied:
yes

vtjnash commented on src/julia_internal.h:
```diff
@@ -859,6 +859,92 @@ jl_method_t *jl_make_opaque_closure_method(jl_module_t *module, jl_value_t *name
     int nargs, jl_value_t *functionloc, jl_code_info_t *ci, int isva, int isinferred);
 JL_DLLEXPORT int jl_is_valid_oc_argtype(jl_tupletype_t *argt, jl_method_t *source);
 
+EXTERN_INLINE_DECLARE enum jl_partition_kind decode_restriction_kind(ptr_kind_union_t pku) JL_NOTSAFEPOINT
```
everything exported should have `jl_` namespacing prefix, or the expmap might cause issues

vtjnash commented on src/support/dtypes.h:
```diff
@@ -123,6 +123,13 @@ typedef intptr_t ssize_t;
+#else
+#define EXTERN_INLINE_DECLARE inline __attribute__ ((visibility("default")))
+#endif
+#define EXTERN_INLINE_DEFINE extern inline JL_DLLEXPORT
```
```suggestion
```

Keno replied:
this is required

vtjnash replied:
okay, search wasn't showing anything

vtjnash commented on stdlib/REPL/test/repl.jl:
```diff
@@ -1210,7 +1210,7 @@ end
 # Issue 39427
 @test occursin("does not exist.", sprint(show, help_result(":=")))
-global some_undef_global
+global some_undef_global::Any
```
```suggestion
global some_undef_global
```

vtjnash commented on test/compiler/effects.jl:
```diff
@@ -376,23 +376,18 @@ end
 # since the inference currently doesn't track "world-age" of global variables
 @eval global_assignment_undefinedyet() = $(GlobalRef(@__MODULE__, :UNDEFINEDYET)) = 42
-setglobal!_nothrow_undefinedyet() = setglobal!(@__MODULE__, :UNDEFINEDYET, 42)
+setglobal!_nothrow_undefinedyet() = setglobal!(@__MODULE__, :UNDEFINEDYET2, 42)
```
```suggestion
setglobal!_nothrow_undefinedyet() = setglobal!(@__MODULE__, :UNDEFINEDYET, 42)
```

vtjnash commented on test/misc.jl:
```diff
@@ -286,7 +286,7 @@ end
 
 # test that they don't introduce global vars
-global v11801, t11801, names_before_timing
+global v11801::Any, t11801::Any, names_before_timing::Any
```
```suggestion
global v11801, t11801, names_before_timing
```
Type declarations being required seems breaking?

Keno replied:
It's required in the full thing, I believe I changed it back for this PR. I can undo the test changes.

vtjnash commented on test/syntax.jl:
```diff
@@ -2645,8 +2645,8 @@ end
-@test_throws ErrorException("cannot assign a value to imported variable Mod.always_undef from module Mod3") Core.eval(Mod3, :(const always_undef = 3))
-@test_throws ErrorException("cannot assign a value to imported variable Mod3.f") Core.eval(Mod3, :(const f = 3))
+@test_throws ErrorException("cannot declare Mod3.always_undef constant; it already has a value") Core.eval(Mod3, :(const always_undef = 3))
+@test_throws ErrorException("cannot declare Mod3.f constant; it already has a value") Core.eval(Mod3, :(const f = 3))
```
Seems a bit inaccurate, since the issue isn't that it has a value (it could be #undef), but that it has a declaration already (being imported)

vtjnash commented on test/staged.jl:
```diff
@@ -311,8 +311,8 @@ end
 end
-@test_throws ErrorException f33243()
-global x33243
+@test_skip(@test_throws ErrorException f33243())
```
```suggestion
@test_throws ErrorException f33243()
```

vtjnash commented on src/codegen.cpp:
```diff
@@ -6408,26 +6474,31 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr, ssize_t ssaidx_
-            }
-            if (bp) {
+                assert(vi.boxroot);
                 Value *mdargs[] = { name, literal_pointer_val(ctx, (jl_value_t*)mod), bp, literal_pointer_val(ctx, bnd) };
```
Are we sure this code path gets tested at all? It seems like `bnd` is always a nullptr, and the arguments look to be in completely the wrong order and of the wrong count

Keno replied:
Does look suspicious. I'll look into it.

Keno replied:
#55532 I think the codepath is dead

vtjnash commented on src/module.c:
```diff
@@ -167,24 +172,35 @@ static jl_globalref_t *jl_new_globalref(jl_module_t *mod, jl_sym_t *name, jl_bin
-    b->usingfailed = 0;
-    b->padding = 0;
     JL_GC_PUSH1(&b);
     b->globalref = jl_new_globalref(mod, name, b);
```
shouldn't this assignment need a jl_gc_wb?

Keno replied:
Probably. In general, the optimization to drop write barriers for freshly allocated objects bothers me, since it's never clear whether somebody simply forgot is explicitly using that property. Let me introduce `jl_gc_wb_fresh` that just expands to nothing, but explicitly annotates this case.

vtjnash replied:
It goes the other way here though, since `b` is old here. It can only be dropped for fields that we see being initially allocated

Keno replied:
Yes, I already added it locally.

vtjnash commented on src/module.c:
```diff
@@ -167,24 +172,35 @@ static jl_globalref_t *jl_new_globalref(jl_module_t *mod, jl_sym_t *name, jl_bin
     return g;
```
no gc_wb should be needed in this function

vtjnash commented on src/module.c:
```diff
@@ -397,19 +466,16 @@ static jl_binding_t *jl_resolve_owner(jl_binding_t *b/*optional*/, jl_module_t *
-            // concurrent import
-            return owner;
-        }
+        jl_atomic_store_release(&bpart->restriction, encode_restriction((jl_value_t*)b2, BINDING_KIND_IMPLICIT));
```
What is making this thread-safe now (since removing cmpswap)?

vtjnash commented on src/module.c:
```diff
@@ -683,14 +783,23 @@ JL_DLLEXPORT void jl_module_public(jl_module_t *from, jl_sym_t *s, int exported)
 
 JL_DLLEXPORT int jl_boundp(jl_module_t *m, jl_sym_t *var, int allow_import) // unlike most queries here, this is currently seq_cst
```
Looks like this isn't being seq-cst anymore, which is required

vtjnash commented on src/module.c:
```diff
@@ -889,39 +989,29 @@ void jl_binding_deprecation_warning(jl_module_t *m, jl_sym_t *s, jl_binding_t *b
+    assert(!jl_bkind_is_some_guard(decode_restriction_kind(pku)) && !jl_bkind_is_some_import(decode_restriction_kind(pku)));
+    if (jl_bkind_is_some_constant(decode_restriction_kind(pku))) {
+        jl_value_t *old = decode_restriction_value(pku);
+        if (rhs == old)
```
jl_egal

vtjnash commented on src/module.c:
```diff
@@ -889,39 +989,29 @@ void jl_binding_deprecation_warning(jl_module_t *m, jl_sym_t *s, jl_binding_t *b
+        jl_value_t *old = decode_restriction_value(pku);
+        if (rhs == old)
+            return NULL;
+        if (jl_typeof(rhs) == jl_typeof(old))
```
```suggestion
        if (jl_typeof(rhs) == jl_typeof(old) || jl_is_type(rhs) || jl_is_module(rhs))
```

vtjnash commented on src/module.c:
```diff
@@ -889,39 +989,29 @@ void jl_binding_deprecation_warning(jl_module_t *m, jl_sym_t *s, jl_binding_t *b
+            jl_errorf("invalid redefinition of constant %s.%s. This redefinition may be permitted using the `const` keyword.",
+                        jl_symbol_name(mod->name), jl_symbol_name(var));
+        else
+            jl_errorf("invalid redefinition of constant %s.%s.",
+                jl_symbol_name(mod->name), jl_symbol_name(var));
```
```suggestion
            jl_safe_printf("WARNING: redefinition of constant %s.%s. This may fail, cause incorrect answers, or produce other errors.\n",
                           jl_symbol_name(mod->name), jl_symbol_name(var));
```

Keno replied:
No, this is the the deliberate change to disallow assignments to constants without the keyword (which no longer goes through this path).

vtjnash replied:
okay, yes, I just lost track of what was using `jl_check_binding_wr`

vtjnash commented on src/module.c:
```diff
@@ -970,14 +1061,20 @@ JL_DLLEXPORT jl_value_t *jl_checked_assignonce(jl_binding_t *b, jl_module_t *mod
+            return;
+        if (!jl_bkind_is_some_guard(decode_restriction_kind(pku))) {
+            jl_errorf("cannot declare %s.%s constant; it already has a value",
+                    jl_symbol_name(b->globalref->mod->name), jl_symbol_name(b->globalref->name));
```
```suggestion
                    jl_symbol_name(mod->name), jl_symbol_name(var));
```

vtjnash commented on src/toplevel.c:
```diff
@@ -310,14 +295,33 @@ static jl_value_t *jl_eval_dot_expr(jl_module_t *m, jl_value_t *x, jl_value_t *f
+                    jl_symbol_name(jl_globalref_mod(b->globalref)->name), jl_symbol_name(jl_globalref_name(b->globalref)));
+        }
+        jl_value_t *old_ty = decode_restriction_value(pku);
+        if (!jl_types_equal(ty, old_ty)) {
```
I don't know why we had an `error` parameter, but looks like it isn't being respected now (for whatever reason it existed)

Keno replied:
Appears to be unused since introduction.

Keno commented:
> Could you add some comments on how the atomic thread-safety of updates are handled right now (i.e. before the final transition to making the partition restrictions field kind be constant)? E.g. what transitions are allowed and what observable effect that has on other concurrent code and inference?

Will add appropriate comments to the header.

Keno committed:
Address some review feedback

Keno committed:
Add jl_ prefix to ptr_kind_union

vtjnash commented on src/jltypes.c:
```diff
@@ -3153,9 +3153,9 @@ void jl_init_types(void) JL_GC_DISABLED
     jl_binding_partition_type->name->atomicfields = binding_partition_atomicfields;
-    const static uint32_t binding_partition_constfields[] = { 0x0002 }; // Set field 2 as constant
+    const static uint32_t binding_partition_constfields[] = { 0x0 };
     jl_binding_partition_type->name->constfields = binding_partition_constfields;
```
```suggestion
```

Keno committed:
Address more review

Keno committed:
Revert some test changes not required for this PR

Keno committed:
Remove unused lpv

Keno committed:
Rebase

Keno committed:
Address final bits of code review

Keno committed:
Fix typo

Keno commented:
I think this should be good to go now. I've gone through and addressed the review comments. I'll run another round of pkgeval. Unless there's something pressing, I think I'll take anything additional post-commit since this is one of those branches that keeps accumulating conflicts and I need to keep the full partition branch up-to-date as well.

Keno committed:
Consider "declared" to be "resolved"

For compat with current semantics.

Keno commented:
@nanosoldier `runtests()`

Keno committed:
Fix check

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/54788#issuecomment-2316698026) failed. Consult the server logs for more details (cc @maleadt).

Keno commented:
@nanosoldier `runtests()`

maleadt commented:
> Consult the server logs for more details (cc @maleadt).

Looks like a compilation failure:

```
Failed to precompile REPLExt [e5eb5ef1-03cf-53a7-ae1d-5a66b08e832b] to "/source/usr/share/julia/compiled/v1.12/REPLExt/jl_6A0kac".
[14458] signal 11 (1): Segmentation fault
in expression starting at /source/usr/share/julia/stdlib/v1.12/REPL/src/precompile.jl:16
ijl_binding_resolved_p at /source/src/module.c:853
ijl_is_binding_deprecated at /source/src/module.c:991
isdeprecated at ./reflection.jl:167 [inlined]
#10 at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPLCompletions.jl:186
filter! at ./array.jl:2927
unknown function (ip: 0x7fceb09967da)
_jl_invoke at /source/src/gf.c:3175 [inlined]
ijl_apply_generic at /source/src/gf.c:3352
append_filtered_mod_names! at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPLCompletions.jl:142
#complete_symbol!#9 at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPLCompletions.jl:185
complete_symbol! at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPLCompletions.jl:154 [inlined]
#complete_identifiers!#55 at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPLCompletions.jl:1214
complete_identifiers! at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPLCompletions.jl:1113 [inlined]
completions at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPLCompletions.jl:1475
#complete_line#89 at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPL.jl:774
complete_line at /source/usr/share/julia/stdlib/v1.12/REPL/src/REPL.jl:771
unknown function (ip: 0x7fceb0995f5d)
_jl_invoke at /source/src/gf.c:3175 [inlined]
ijl_apply_generic at /source/src/gf.c:3352
check_for_hint at /source/usr/share/julia/stdlib/v1.12/REPL/src/LineEdit.jl:385
#143 at /source/usr/share/julia/stdlib/v1.12/REPL/src/LineEdit.jl:2538
_jl_invoke at /source/src/gf.c:3175 [inlined]
ijl_apply_generic at /source/src/gf.c:3352
jl_apply at /source/src/julia.h:2217 [inlined]
jl_f__call_latest at /source/src/builtins.c:875
#invokelatest#2 at ./essentials.jl:1048 [inlined]
invokelatest at ./essentials.jl:1045 [inlined]
#30 at /source/usr/share/julia/stdlib/v1.12/REPL/src/LineEdit.jl:1722
unknown function (ip: 0x7fceb091fcb9)
_jl_invoke at /source/src/gf.c:3175 [inlined]
ijl_apply_generic at /source/src/gf.c:3352
macro expansion at /source/usr/share/julia/stdlib/v1.12/REPL/src/LineEdit.jl:2872 [inlined]
macro expansion at ./lock.jl:273 [inlined]
#282 at /source/usr/share/julia/stdlib/v1.12/REPL/src/LineEdit.jl:2862
unknown function (ip: 0x7fceb091c998)
_jl_invoke at /source/src/gf.c:3175 [inlined]
ijl_apply_generic at /source/src/gf.c:3352
jl_apply at /source/src/julia.h:2217 [inlined]
start_task at /source/src/task.c:1294
Allocations: 35464420 (Pool: 35463648; Big: 772); GC: 27
```

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54788#issuecomment-2316738219) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/593a8d2_vs_16697f3/report.html) is available.

nsajko commented:
This seems to break CI consistently: #55642
